File: backends/shared-backend-configuration/src/main/java/db/migration/common/V41_1__set_salt_in_existing_user_credentials.java
package db.migration.common;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.UUID;

import org.springframework.jdbc.core.JdbcTemplate;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import org.cloudfoundry.credhub.CryptSaltFactory;

import org.cloudfoundry.credhub.utils.UuidUtil;
import org.flywaydb.core.api.migration.spring.SpringJdbcMigration;

import static java.nio.charset.StandardCharsets.UTF_8;

@SuppressWarnings("unused")
public class V41_1__set_salt_in_existing_user_credentials implements SpringJdbcMigration {

  @SuppressFBWarnings(
    value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE",
    justification = "The database will definitely exist"
  )
  @Override
  public void migrate(final JdbcTemplate jdbcTemplate) throws Exception {

    final String databaseName = jdbcTemplate
      .getDataSource()
      .getConnection()
      .getMetaData()
      .getDatabaseProductName()
      .toLowerCase();

    final CryptSaltFactory saltFactory = new CryptSaltFactory();

    final List<UUID> uuids = jdbcTemplate.query("select uuid from user_credential", (rowSet, rowNum) -> {
      final byte[] uuidBytes = rowSet.getBytes("uuid");
      if ("postgresql".equals(databaseName)) {
        return UUID.fromString(new String(uuidBytes, UTF_8));
      } else {
        final ByteBuffer byteBuffer = ByteBuffer.wrap(uuidBytes);
        return new UUID(byteBuffer.getLong(), byteBuffer.getLong());
      }
    });

    for (final UUID uuid: uuids) {
      final String salt = saltFactory.generateSalt();

      jdbcTemplate.update(
          "update user_credential set salt = ? where uuid = ?",
          new Object[]{salt, getUuidParam(databaseName, uuid)}
      );
    }
  }

  private Object getUuidParam(final String databaseName, final UUID uuid) {
    if ("postgresql".equals(databaseName)) {
      return uuid;
    } else {
      return UuidUtil.uuidToByteArray(uuid);
    }
  }
}

File: backends/shared-backend-configuration/src/main/kotlin/db/migration/common/V41_1__set_salt_in_existing_user_credentials.kt
package db.migration.common

import org.cloudfoundry.credhub.CryptSaltFactory
import org.cloudfoundry.credhub.utils.UuidUtil
import org.flywaydb.core.api.migration.BaseJavaMigration
import org.flywaydb.core.api.migration.Context
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.datasource.SingleConnectionDataSource
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.sql.ResultSet
import java.util.UUID

class V41_1__set_salt_in_existing_user_credentials : BaseJavaMigration() {
    @Throws(Exception::class)
    override fun migrate(context: Context) {
        val jdbcTemplate = JdbcTemplate(SingleConnectionDataSource(context.connection, true))
        val databaseName = jdbcTemplate
            .dataSource
            ?.getConnection()
            ?.metaData
            ?.databaseProductName
            ?.lowercase()
        val saltFactory = CryptSaltFactory()
        val uuids = jdbcTemplate.query("select uuid from user_credential") { rowSet: ResultSet, _: Int ->
            val uuidBytes = rowSet.getBytes("uuid")
            if ("postgresql" == databaseName) {
                return@query UUID.fromString(String(uuidBytes, StandardCharsets.UTF_8))
            } else {
                val byteBuffer = ByteBuffer.wrap(uuidBytes)
                return@query UUID(byteBuffer.long, byteBuffer.long)
            }
        }
        for (uuid in uuids) {
            val salt = saltFactory.generateSalt()
            jdbcTemplate.update(
                "update user_credential set salt = ? where uuid = ?",
                *arrayOf(salt, databaseName?.let { getUuidParam(it, uuid) })
            )
        }
    }

    private fun getUuidParam(databaseName: String, uuid: UUID): Any {
        return if ("postgresql" == databaseName) {
            uuid
        } else {
            UuidUtil.uuidToByteArray(uuid)
        }
    }
}
